core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -98,22 +112,57 @@ public void processExpiration() {
                      expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, ice.getCreated(), currentTimeMillis);
                      expiredTransient = ExpiryHelper.isExpiredTransient(maxIdle, ice.getLastUsed(), currentTimeMillis);
                   }
-                  // We check lifespan first as this is much less expensive to remove than max idle.
-                  if (expiredMortal) {
-                     addAndWaitIfFull(handleLifespanExpireEntry(ice.getKey(), value, lifespan, false), futures);
-                  } else if (expiredTransient) {
-                     addAndWaitIfFull(actualRemoveMaxIdleExpireEntry(ice.getKey(), value, maxIdle, false), futures);
+                  if (expiredMortal || expiredTransient) {
+                     if (expirationPermits.tryAcquire(1, TimeUnit.MINUTES)) {
+                        removedEntries++;
+                        CompletableFuture<?> stage;
+                        // We check lifespan first as this is much less expensive to remove than max idle.
+                        if (expiredMortal) {
+                           stage = handleLifespanExpireEntry(ice.getKey(), value, lifespan, false);
+                        } else {
+                           stage = actualRemoveMaxIdleExpireEntry(ice.getKey(), value, maxIdle, false);
+                        }
+                        futures.add(stage);
+                        stage.whenComplete((t, obj) -> {
+                           if (trace) {
+                              log.tracef("There was a problem expiring the entry for key: " + ice.getKey(), t);
```

Comment:
For the final futures we call `join()` and we log a warning, we should do the same here.

Reply:
Oops, this was supposed to be `if (trace && t !=null)` as well :)

Reply:
Fixed